File: common/src/main/java/org/valkyrienskies/clockwork/util/render/WingBlockEntityRenderer.java
package org.valkyrienskies.clockwork.util.render;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import com.simibubi.create.foundation.render.CachedBufferer;
import com.simibubi.create.foundation.render.SuperByteBuffer;
import com.simibubi.create.foundation.tileEntity.renderer.SmartTileEntityRenderer;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.core.Direction;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import org.valkyrienskies.clockwork.ClockWorkPartials;
import org.valkyrienskies.clockwork.content.materials.solids.colorblock.ColorBlockEntity;
import org.valkyrienskies.clockwork.util.blocktype.ConnectedWingAlike;

public class WingBlockEntityRenderer extends SmartTileEntityRenderer<ColorBlockEntity> {
    public WingBlockEntityRenderer(BlockEntityRendererProvider.Context context) {
        super(context);
    }

    @Override
    protected void renderSafe(ColorBlockEntity be, float partialTicks, PoseStack ms, MultiBufferSource buffer, int light, int overlay) {
        super.renderSafe(be, partialTicks, ms, buffer, light, overlay);
        VertexConsumer vb = buffer.getBuffer(RenderType.cutout());
        BlockState state = be.getBlockState();
        int color = be.getColor();
        Direction facing = state.getValue(BlockStateProperties.FACING);

        SuperByteBuffer middle = applyColor(color, CachedBufferer.partial(ClockWorkPartials.WING_MIDDLE, state));
        SuperByteBuffer side = CachedBufferer.partial(ClockWorkPartials.WING_SIDE, state);
        SuperByteBuffer side_vertical = CachedBufferer.partial(ClockWorkPartials.WING_SIDE_VERTICAL, state);

        switch (facing) {
            case NORTH, SOUTH -> middle.rotateCentered(Direction.EAST, (float) Math.toRadians(90)).light(light).renderInto(ms, vb);
            case EAST, WEST -> middle.rotateCentered(Direction.NORTH, (float) Math.toRadians(90)).light(light).renderInto(ms, vb);
            default -> middle.light(light).renderInto(ms, vb);
        }

        if (state.getValue(BlockStateProperties.NORTH)) {
            switch (facing) {
                case EAST, WEST -> applyColor(color, side_vertical)
                        .light(light).renderInto(ms, vb);
                case UP, DOWN -> applyColor(color, side)
                        .light(light).renderInto(ms, vb);
            }
        }

        if (state.getValue(BlockStateProperties.SOUTH)) {
            switch (facing) {
                case EAST, WEST -> applyColor(color, side_vertical)
                        .rotateCentered(Direction.UP, (float) Math.toRadians(180))
                        .light(light).renderInto(ms, vb);
                case UP, DOWN -> applyColor(color, side)
                        .rotateCentered(Direction.UP, (float) Math.toRadians(180))
                        .light(light).renderInto(ms, vb);
            }
        }

        if (state.getValue(BlockStateProperties.EAST)) {
            switch (facing) {
                case NORTH, SOUTH -> applyColor(color, side_vertical)
                        .rotateCentered(Direction.UP, (float) Math.toRadians(270))
                        .light(light).renderInto(ms, vb);
                case UP, DOWN -> applyColor(color, side)
                        .rotateCentered(Direction.UP, (float) Math.toRadians(270))
                        .light(light).renderInto(ms, vb);
            }
        }

        if (state.getValue(BlockStateProperties.WEST)) {
            switch (facing) {
                case NORTH, SOUTH -> applyColor(color, side_vertical)
                        .rotateCentered(Direction.UP, (float) Math.toRadians(90))
                        .light(light).renderInto(ms, vb);
                case UP, DOWN -> applyColor(color, side)
                        .rotateCentered(Direction.UP, (float) Math.toRadians(90))
                        .light(light).renderInto(ms, vb);
            }
        }

        if (state.getValue(BlockStateProperties.UP)) {
            switch (facing) {
                case NORTH, SOUTH -> applyColor(color, side)
                        .rotateCentered(Direction.EAST, (float) Math.toRadians(90))
                        .light(light).renderInto(ms, vb);
                case EAST, WEST -> applyColor(color, side)
                        .rotateCentered(Direction.EAST, (float) Math.toRadians(90))
                        .rotateCentered(Direction.NORTH, (float) Math.toRadians(270))
                        .light(light).renderInto(ms, vb);
            }
        }

        if (state.getValue(BlockStateProperties.DOWN)) {
            switch (facing) {
                case NORTH, SOUTH -> applyColor(color, side)
                        .rotateCentered(Direction.EAST, (float) Math.toRadians(270))
                        .light(light).renderInto(ms, vb);
                case EAST, WEST -> applyColor(color, side)
                        .rotateCentered(Direction.NORTH, (float) Math.toRadians(90))
                        .rotateCentered(Direction.UP, (float) Math.toRadians(270))
                        .light(light).renderInto(ms, vb);
            }
        }
    }

    private SuperByteBuffer applyColor(int color, SuperByteBuffer buf) {
        if (color != -1)
            buf.color(color);
        return buf;
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/client/render/WingBlockEntityRenderer.kt
package org.valkyrienskies.clockwork.client.render

import com.mojang.blaze3d.vertex.PoseStack
import com.simibubi.create.foundation.blockEntity.renderer.SmartBlockEntityRenderer
import com.simibubi.create.foundation.render.CachedBufferer
import com.simibubi.create.foundation.render.SuperByteBuffer
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.core.Direction
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import org.valkyrienskies.clockwork.ClockworkPartials
import org.valkyrienskies.clockwork.content.generic.ColorBlockEntity

class WingBlockEntityRenderer(context: BlockEntityRendererProvider.Context?) :
    SmartBlockEntityRenderer<ColorBlockEntity>(context) {
    override fun renderSafe(
            be: ColorBlockEntity,
            partialTicks: Float,
            ms: PoseStack,
            buffer: MultiBufferSource,
            light: Int,
            overlay: Int
    ) {
        super.renderSafe(be, partialTicks, ms, buffer, light, overlay)
        val vb = buffer.getBuffer(RenderType.cutout())
        val state: BlockState = be.blockState
        val color: Int = be.getColor()
        val facing = state.getValue(BlockStateProperties.FACING)
        val middle = applyColor(color, CachedBufferer.partial(ClockworkPartials.WING_MIDDLE, state))
        val side = CachedBufferer.partial(ClockworkPartials.WING_SIDE, state)
        val sideVertical = CachedBufferer.partial(ClockworkPartials.WING_SIDE_VERTICAL, state)
        when (facing) {
            Direction.NORTH, Direction.SOUTH -> middle.rotateCentered(Direction.EAST, Math.toRadians(90.0).toFloat())
                .light(light).renderInto(ms, vb)

            Direction.EAST, Direction.WEST -> middle.rotateCentered(Direction.NORTH, Math.toRadians(90.0).toFloat())
                .light(light).renderInto(ms, vb)

            else -> middle.light(light).renderInto(ms, vb)
        }
        if (state.getValue(BlockStateProperties.NORTH)) {
            when (facing) {
                Direction.EAST, Direction.WEST -> applyColor(color, sideVertical)
                    .light(light).renderInto(ms, vb)

                Direction.UP, Direction.DOWN -> applyColor(color, side)
                    .light(light).renderInto(ms, vb)

                else -> {}
            }
        }
        if (state.getValue(BlockStateProperties.SOUTH)) {
            when (facing) {
                Direction.EAST, Direction.WEST -> applyColor(color, sideVertical)
                    .rotateCentered(Direction.UP, Math.toRadians(180.0).toFloat())
                    .light(light).renderInto(ms, vb)

                Direction.UP, Direction.DOWN -> applyColor(color, side)
                    .rotateCentered(Direction.UP, Math.toRadians(180.0).toFloat())
                    .light(light).renderInto(ms, vb)

                else -> {}
            }
        }
        if (state.getValue(BlockStateProperties.EAST)) {
            when (facing) {
                Direction.NORTH, Direction.SOUTH -> applyColor(color, sideVertical)
                    .rotateCentered(Direction.UP, Math.toRadians(270.0).toFloat())
                    .light(light).renderInto(ms, vb)

                Direction.UP, Direction.DOWN -> applyColor(color, side)
                    .rotateCentered(Direction.UP, Math.toRadians(270.0).toFloat())
                    .light(light).renderInto(ms, vb)

                else -> {}
            }
        }
        if (state.getValue(BlockStateProperties.WEST)) {
            when (facing) {
                Direction.NORTH, Direction.SOUTH -> applyColor(color, sideVertical)
                    .rotateCentered(Direction.UP, Math.toRadians(90.0).toFloat())
                    .light(light).renderInto(ms, vb)

                Direction.UP, Direction.DOWN -> applyColor(color, side)
                    .rotateCentered(Direction.UP, Math.toRadians(90.0).toFloat())
                    .light(light).renderInto(ms, vb)

                else -> {}
            }
        }
        if (state.getValue(BlockStateProperties.UP)) {
            when (facing) {
                Direction.NORTH, Direction.SOUTH -> applyColor(color, side)
                    .rotateCentered(Direction.EAST, Math.toRadians(90.0).toFloat())
                    .light(light).renderInto(ms, vb)

                Direction.EAST, Direction.WEST -> applyColor(color, side)
                    .rotateCentered(Direction.EAST, Math.toRadians(90.0).toFloat())
                    .rotateCentered(Direction.NORTH, Math.toRadians(270.0).toFloat())
                    .light(light).renderInto(ms, vb)

                else -> {}
            }
        }
        if (state.getValue(BlockStateProperties.DOWN)) {
            when (facing) {
                Direction.NORTH, Direction.SOUTH -> applyColor(color, side)
                    .rotateCentered(Direction.EAST, Math.toRadians(270.0).toFloat())
                    .light(light).renderInto(ms, vb)

                Direction.EAST, Direction.WEST -> applyColor(color, side)
                    .rotateCentered(Direction.NORTH, Math.toRadians(90.0).toFloat())
                    .rotateCentered(Direction.UP, Math.toRadians(270.0).toFloat())
                    .light(light).renderInto(ms, vb)

                else -> {}
            }
        }
    }

    private fun applyColor(color: Int, buf: SuperByteBuffer): SuperByteBuffer {
        if (color != -1) buf.color(color)
        return buf
    }
}
